### PR TITLE
[ENH] Setup performance benchmarks using asv

### LINKED ISSUE #4573: Set up a benchmark suite to assess performance and check for improvment or regression
This kind of fix makes me think it'd be good to have a bunch of benchmark suite to assess performance and check for improvment or regression.

@htwangtw I think you had a look into this, no?
- https://github.com/nilearn/nilearn/issues/3313

By the way @htwangtw if you have time for a review of this one: it is a small one.

_Originally posted by @Remi-Gau in https://github.com/nilearn/nilearn/issues/4571#issuecomment-2379663727_

### PR DESCRIPTION
**Pull request by man-shu**
- Closes #4573

Changes proposed in this pull request:

- add minimal benchmarks (more to follow in subsequent PRs)
- `asv.conf.json` has comments in it which annoys precommit's json, so ignored it

TODO
- [x] a github workflow to run the benchmarks and push the HTML report somewhere (daily and on commit triggers)
- [x] add a doc page/section on how to run benchmarks locally

### CODE REVIEW

**Comment by Remi-Gau** [comment]
> * `asv.conf.json` has comments in it which annoys precommit's json, so ignored it

yeah that's because comments are not allowed in json: not sure what kind of devilry this is

**Review by bthirion — COMMENTED** [review]
LGTM so far. Will need dedicated explanation in the doc too.

**Comment by man-shu** [comment]
> @man-shu I'd that for this PR, it'd be good to at least have minimal doc on "how to run it locally": this way we can check this for ourselves as part of the PR review process

Yep I agree. Added it to the TODO

**Comment by man-shu** [comment]
Ok I was wrong by a factor of 10 🤣

**Comment by Remi-Gau** [comment]
> Ok I was wrong by a factor of 10 🤣

so 0.2 hours? 😉

**Comment by man-shu** [comment]
> so 0.2 hours? 😉

🤣

It's running here now: https://github.com/man-shu/nilearn_benchmarks/actions/runs/14090132382/job/39464456389
Should take ~3 hours

**Comment by man-shu** [comment]
Some of the benchmarks here can be seen on our website now: https://nilearn.github.io/benchmarks/

**Comment by man-shu** [comment]
Some examples to see:

1. under `comparisons`, we have comparisons of an operation (say, slicing) between implementation/image loading method. 
   - first loading the image via nibabel.load and then slicing is [much faster](https://nilearn.github.io/benchmarks/#comparisons.image.CompareSlice.time_compare_slice?x-axis=loader) and [memory-efficient](https://nilearn.github.io/benchmarks/#comparisons.image.CompareSlice.peakmem_compare_slice?x-axis=loader) than loading via nilearn.load_img and then slicing.
2. under `tracking`, we are tracking the performance of (say, NiftiMasker with different parameter values) over commits:
   - when we set smoothing_fwhm to 6, NiftiMasker takes [more time](https://nilearn.github.io/benchmarks/#tracking.maskers.NiftiMaskerBenchmark.time_nifti_masker) and [memory](https://nilearn.github.io/benchmarks/#tracking.maskers.NiftiMaskerBenchmark.peakmem_nifti_masker) than when it's None, while `detrend` makes no difference
3. we can also see recent increases in the benchmark list and clicking on them would take us to the plot and clicking on the commit would take us to the diff: https://nilearn.github.io/benchmarks/#/summarylist

**Comment by bthirion** [comment]
> Some examples to see:
> 
>     1. under `comparisons`, we have comparisons of an operation (say, slicing) between implementation/image loading method.
>        
>        * first loading the image via nibabel.load and then slicing is [much faster](https://nilearn.github.io/benchmarks/#comparisons.image.CompareSlice.time_compare_slice?x-axis=loader) and [memory-efficient](https://nilearn.github.io/benchmarks/#comparisons.image.CompareSlice.peakmem_compare_slice?x-axis=loader) than loading via nilearn.load_img and then slicing.
> 
>     2. under `tracking`, we are tracking the performance of (say, NiftiMasker with different parameter values) over commits:
>        
>        * when we set smoothing_fwhm to 6, NiftiMasker takes [more time](https://nilearn.github.io/benchmarks/#tracking.maskers.NiftiMaskerBenchmark.time_nifti_masker) and [memory](https://nilearn.github.io/benchmarks/#tracking.maskers.NiftiMaskerBenchmark.peakmem_nifti_masker) than when it's None, while `detrend` makes no difference
> 
>     3. we can also see recent increases in the benchmark list and clicking on them would take us to the plot and clicking on the commit would take us to the diff: https://nilearn.github.io/benchmarks/#/summarylist

Thx ! Is there a way to get these explantions when looking at the figures, e.g. in a comment, legend etc. ATM I can't see it.

**Comment by man-shu** [comment]
The legends are there in cases where we are changing the parameters, like in the top left corner here https://nilearn.github.io/benchmarks/#tracking.maskers.NiftiMaskerBenchmark.time_nifti_masker:

Screenshot 2025-03-28 at 12 07 21 PM

In addition, the code for a given benchmark can be seen when you hover over the respective plot on this page: https://nilearn.github.io/benchmarks/

Screenshot 2025-03-28 at 12 04 28 PM

File: asv_benchmarks/benchmarks/comparisons/image.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        img = load(loader)[1]
+        mean_img(img)
+
+    def peakmem_compare_mean(self, loader):
+        """Peak memory of loading followed by taking the mean."""
+        img = load(loader)[1]
+        mean_img(img)
+
+
+class CompareSlice(Benchmark):
```
Comment (man-shu) [inline asv_benchmarks/benchmarks/comparisons/image.py:51]: I think we are pretty much stuck with the hovering thing on the webpages.
However, you could look at the code here if needed...

File: asv_benchmarks/asv.conf.json [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    "matrix": {
+        "numpy": ["1.22.4"],
+        "scipy": ["1.8.0"],
+        "joblib": ["1.2.0"],
+        "pandas": ["2.2.0"],
+        "scikit-learn": ["1.4.0"],
+        "nibabel": ["5.2.0"],
+    }
+
+    // Combinations of libraries/python versions can be excluded/included
```
Comment (bthirion) [inline asv_benchmarks/asv.conf.json:83]: Are these lines to be removed ?

Comment (man-shu) [reply asv_benchmarks/asv.conf.json:83]: Those come from the default config generated by asv. I think we should keep them if we want to use some that functionality later on.

File: asv_benchmarks/benchmarks/comparisons/image.py (hunk 1 above)
Comment (man-shu) [reply asv_benchmarks/benchmarks/comparisons/image.py:51]: Actually, all the code excerpts are there in a JSON file.
It is definitely possible to write a script to add more views/plots etc. on the benchmark webpages. 
But I would keep that as a long-term-nice-to-have.

**Comment by man-shu** [comment]
So here's a section on the benchmarks: https://output.circle-artifacts.com/output/job/d335787f-1f86-4639-97ab-96fa20b7d3be/artifacts/0/dev/development.html#performance-monitoring

and the description for the workflow: https://output.circle-artifacts.com/output/job/d335787f-1f86-4639-97ab-96fa20b7d3be/artifacts/0/dev/ci.html#benchmark-yml. 
I tested it on the old repo for benchmarks, but we could know for sure after merging this PR.

File: .github/workflows/benchmark.yml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                SSH_AUTH_SOCK: /tmp/ssh_agent_nilearn_github_io.sock
+            run: |
+                git clone --depth 1 git@github.com:nilearn/nilearn.github.io.git --recursive
+                cd nilearn.github.io
+                git config --global user.name "GitHub Actions"
+                git config --global user.email "actions@github.com"
+                git submodule update --remote --merge
+                git add benchmarks
+                git commit -m "Deploy benchmarks"
+                git push origin main
```
Comment (Remi-Gau) [inline .github/workflows/benchmark.yml:135]: quick suggestion:
- let's not submodule things in the doc repo
- let's deploy directly to github pages from the benchmark repo
- in both cases the URL will be: https://nilearn.github.io/benchmarks/
- this way we avoid having a submodule and keep it updated (will also make the history of the doc repo shorter: https://github.com/nilearn/nilearn.github.io/commits/main/)

File: asv_benchmarks/asv.conf.json [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    "repo": "..",
+
+    // The Python project's subdirectory in your repo.  If missing or
+    // the empty string, the project is assumed to be located at the root
+    // of the repository.
+    // "repo_subdir": "",
+
+    // Customizable commands for building, installing, and
+    // uninstalling the project. See asv.conf.json documentation.
+    "install_command": ["python -mpip install {wheel_file}"],
```
Comment (Remi-Gau) [inline asv_benchmarks/asv.conf.json:23]:
```suggestion
    "install_command": ["python -m pip install {wheel_file}"],
```
isn't a space required here?

File: .github/workflows/benchmark.yml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    # everday at 10pm UTC
+    -   cron: 0 22 * * *
+
+jobs:
+    benchmark:
+        if: >-
+            (contains(github.event.head_commit.message, '[bm]') ||
+            github.event_name == 'schedule') &&
+            github.repository == 'nilearn/nilearn'
+        runs-on: ubuntu-22.04
```
Comment (Remi-Gau) [inline .github/workflows/benchmark.yml:58]:
```suggestion
        # TODO bump to 24.04 when bumping minimum python to 3.10
        runs-on: ubuntu-22.04
```

File: asv_benchmarks/asv.conf.json (hunk 2 above)
Comment (man-shu) [reply asv_benchmarks/asv.conf.json:83]: Yep, for now, this config is only being read by asv and none of our own scripts, so we don't need any additional dependencies.

File: .github/workflows/benchmark.yml (hunk 3 above)
Comment (man-shu) [reply .github/workflows/benchmark.yml:135]: > in both cases the URL will be: https://nilearn.github.io/benchmarks/

I see. I did not know this. Let's try that!

File: asv_benchmarks/asv.conf.json (hunk 4 above)
Comment (man-shu) [reply asv_benchmarks/asv.conf.json:23]: The default install command from asv wasn't working, so copied this from sklearn: https://github.com/scikit-learn/scikit-learn/blob/a867ed7cc6f86146ce2a790531dbbcfb9cafe318/asv_benchmarks/asv.conf.json#L18

But let me try adding the space.

File: .github/workflows/benchmark.yml [hunk 6]
```diff
@@ -0,0 +1,108 @@
+---
+# TL;DR: this workflow runs the benchmarks on the latest commit on the main
+# and then deploys the updated benchmarks to
+# `nilearn.github.io/benchmarks <https://nilearn.github.io/benchmarks/>`_.
+# You can trigger it manually by adding ``[bm]`` to the commit message on the
+# main. It will also run on a schedule every day at 10pm UTC.
```
Comment (bthirion) [inline .github/workflows/benchmark.yml:6]: Every day ? sounds a bit heavy.

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -925,6 +925,65 @@ returned by the ``request_mocker`` pytest fixture, defined in
...
+      pip install asv
+
+Then, change to the ``asv_benchmarks`` directory:
+
+.. code-block:: bash
+
+      cd asv_benchmarks
+
+To run a specific benchmark on the current HEAD of your clone of the
+repository, use the following command:
```
Comment (bthirion) [inline CONTRIBUTING.rst:960]: Maybe explain where we find all the available benchmarks.

**Review by bthirion — COMMENTED** [review]
I have only minor comments. LGTM overall.

File: .github/workflows/benchmark.yml (hunk 6 above)
Comment (emdupre) [reply .github/workflows/benchmark.yml:6]: Strong +1. I think @Remi-Gau [made the same point](https://github.com/nilearn/nilearn/issues/4573#issuecomment-2750913464) in the linked issue.

File: .github/workflows/benchmark.yml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+#
+# Step-wise details:
+#
+# - The benchmarks are set to run everday at 10pm UTC via cron.
```
Comment (Remi-Gau) [inline .github/workflows/benchmark.yml:10]:
```suggestion
# - The benchmarks are set to run at regular intervals via cron.
```

File: .github/workflows/benchmark.yml [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+###
+name: Run and deploy Nilearn benchmarks
+
+on:
+    push:
+        branches:
+        -   main
+    schedule:
+    # everday at 10pm UTC
+    -   cron: 0 22 * * *
```
Comment (Remi-Gau) [inline .github/workflows/benchmark.yml:44]: I think weekly would be more than enough, no?

File: .github/workflows/benchmark.yml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+jobs:
+    benchmark:
+        if: >-
+            (contains(github.event.head_commit.message, '[bm]') ||
```
Comment (Remi-Gau) [inline .github/workflows/benchmark.yml:49]: TODO
we'll have to test the trigger via commit message after merge

File: .github/workflows/benchmark.yml [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            run: |
+                cd benchmarks_repo
+                cp -r ../results .
+                cp -r ../html/* .
+                git config --global user.name "GitHub Actions"
+                git config --global user.email "actions@github.com"
+                git add .
+                git commit -m "Update benchmark results and HTML"
+                git push origin main
+                cd ..
```
Comment (Remi-Gau) [inline .github/workflows/benchmark.yml:108]: are we sure we always want to do this?

say we trigger the workflow on a PR, then we would update our benchmark results repo with results from a PR.

I think we want the benchmark repo to only have things from the main branch, no?

File: CONTRIBUTING.rst [hunk 12]
```diff
@@ -925,6 +925,65 @@ returned by the ``request_mocker`` pytest fixture, defined in
 ``Sender`` class it contains provide information on how to write a test using
 this fixture. Existing tests can also serve as examples.
 
+Performance monitoring
+----------------------
+
+As of March 2025, we are starting to incorporate performance monitoring in the
```
Comment (Remi-Gau) [inline CONTRIBUTING.rst:931]:
```suggestion
Nilearn includes performance monitoring in the
```

I would keep it more vague

File: .github/workflows/benchmark.yml (hunk 11 above)
Comment (man-shu) [reply .github/workflows/benchmark.yml:108]: As of now, this workflow is only for the main, it can't be triggered from PR. 
We decided to hold off on running benchmarks from PRs.

If someone wants to compare with their work on a PR it would have to be done locally. I could add more info on that in the doc but I haven't tried that myself yet.

File: .github/workflows/benchmark.yml (hunk 6 above)
Comment (man-shu) [reply .github/workflows/benchmark.yml]: Ok let's do weekly then.

File: CONTRIBUTING.rst (hunk 7 above)
Comment (man-shu) [reply CONTRIBUTING.rst:960]: I mentioned that in L941. Could also add link

File: .github/workflows/benchmark.yml (hunk 11 above)
Comment (Remi-Gau) [reply .github/workflows/benchmark.yml:108]: ha yeah but then it feels weird to have the workflow "triggerable" by commit message on main

ok let's keep thing as is for now and then get back to this when we allow to run this on PR

Comment (man-shu) [reply .github/workflows/benchmark.yml:108]: Oh, you mean one could potentially name their PR like "[bm] ...." and that would trigger this (since we squash our PR commits). I did not think about that tbh.

I was simply thinking of doing an empty commit with [bm] on the main to trigger it...

File: asv_benchmarks/benchmarks/tracking/maskers.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class NiftiMaskerBenchmark(Benchmark):
+    """
+    Benchmark for applying a mask to an image using NiftiMasker with different
+    parameters.
+    """
+
+    # try different combinations of parameters for the NiftiMasker object
+    param_names = ["smoothing_fwhm", "detrend"]
```
Comment (htwangtw) [inline asv_benchmarks/benchmarks/tracking/maskers.py:16]: Out of curiosity: Is there a reason for picking these two specific parameters only?

Comment (man-shu) [reply asv_benchmarks/benchmarks/tracking/maskers.py:16]: `detrend` is arbitrary just to test the parameterizing in asv

I did want to look into smoothing though because of this issue: https://neurostars.org/t/problem-with-smoothing-in-first-level-analysis/29553/2

But that's probably GLM-specific.

Comment (htwangtw) [reply asv_benchmarks/benchmarks/tracking/maskers.py:16]: Yeah I think that is a good reason to include smoothing here.

**Comment by man-shu** [comment]
Took some inspiration from https://github.com/scikit-learn/scikit-learn-benchmarks and reorganized + added logs in our https://github.com/nilearn/benchmarks.

I think this is ready for merge.

**Review by htwangtw — APPROVED** [review]
LGTM!
